### ISSUE TITLE
high load with firefox 52.9.0esr 32bit on 64bit linuxmint

### ISSUE THREAD

**Issue by sh-dvl** [uBlockOrigin/uBlock-issues#150]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [x] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

fresh firefox 52.9.0esr 32bit from mozilla on a nearly fresh linuxmint 18.3 64bit with 4GB RAM

starting firefox leads immediatly to high load and unusable firefox.

same profile with 64bit-firefox: no problem (but firefox needs more memory due to 64bit)

so maybe a bitsize-issue?

### A specific URL where the issue occurs

any

### Steps to Reproduce

1. [First Step]
install 32bit firefox 52.9.0esr
2. [Second Step]
install ublock-origin
3. [and so on...]
start firefox, try to select menus or try to load a website, do anything and wait forever...

### Expected behavior:

just works

### Actual behavior:

up to 100% load (top) without any progress visible

### Your environment

* uBlock Origin version: 1.16.14, but also tried earlier ones
* Browser Name and version: firefox 52.9.0esr 32bit
* Operating System and version: linuxmint 18.3 64bit xfce-gui
* System: i5-6200U with 4GB, at least 2GB available during tests _after_ starting firefox (no out-of-memory-issue)

**Comment by gwarser** [comment]
You know uBO is not native application? Why you think it's not Firefox fault? Does this happen if you disable and enable uBO from about:addons page? If only on startup - try starting Firefox with `-jsconsole` command parameter and see if something is logged. Try recording performance profile https://perf-html.io/ will be hard to do this on startup, but if it's uBO then should also happen when you disable - enable from about:addons.

**Comment by uBlock-user** [comment]
> same profile with 64bit-firefox: no problem

So a bad 32 bit firefox build possibly

**Comment by gwarser** [comment]
ESR 32 bits, uBO legacy 1.16.4.4:
```
Error: Script terminated by timeout at:
HNTrieBuilder.prototype.vacuum@chrome://ublock0/content/js/hntrie.js:323:9
HNTrieBuilder.fromDomainOpt@chrome://ublock0/content/js/hntrie.js:85:12
FilterOriginHitSet.prototype<.matchOrigin.value@chrome://ublock0/content/js/static-net-filtering.js:890:30
FilterOrigin.prototype.match@chrome://ublock0/content/js/static-net-filtering.js:794:12
FilterBucket.prototype.match@chrome://ublock0/content/js/static-net-filtering.js:1309:14
FilterContainer.prototype.matchTokens@chrome://ublock0/content/js/static-net-filtering.js:2545:29
FilterContainer.prototype.matchStringGenericHide@chrome://ublock0/content/js/static-net-filtering.js:2573:21
FilterContainer.prototype.matchStringExactType@chrome://ublock0/content/js/static-net-filtering.js:2597:16
PageStore.prototype.init@chrome://ublock0/content/js/pagestore.js:322:22
PageStore@chrome://ublock0/content/js/pagestore.js:244:5
PageStore.factory@chrome://ublock0/content/js/pagestore.js:256:17
µb.bindTabToPageStats@chrome://ublock0/content/js/tab.js:857:21
vAPI.tabs.onNavigation@chrome://ublock0/content/js/tab.js:493:21
vAPI.net.registerListeners/locationChangedListener@chrome://ublock0/content/js/vapi-background.js:2402:9
  hntrie.js:323:9
```

Hmmm... Looks like you may be right.

**Comment by gwarser** [comment]
1.16.14 is worse, ~15x:

```
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIDOMWindowUtils.isParentWindowMainWidgetVisible]  nsPrompter.js:350
Error: Script terminated by timeout at:
HNTrieBuilder.prototype.vacuum@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/hntrie.js:333:13
HNTrieBuilder.fromDomainOpt@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/hntrie.js:85:12
FilterOriginHitSet.prototype<.matchOrigin.value@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/static-net-filtering.js:890:30
FilterOrigin.prototype.match@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/static-net-filtering.js:794:12
FilterBucket.prototype.match@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/static-net-filtering.js:1319:14
FilterContainer.prototype.matchTokens@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/static-net-filtering.js:2536:29
FilterContainer.prototype.matchStringGenericHide@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/static-net-filtering.js:2564:21
FilterContainer.prototype.matchStringExactType@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/static-net-filtering.js:2588:16
PageStore.prototype.init@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/pagestore.js:322:22
PageStore@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/pagestore.js:244:5
PageStore.factory@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/pagestore.js:256:17
µb.bindTabToPageStats@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/tab.js:857:21
vAPI.tabs.onUpdated@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/tab.js:514:5
vAPI.tabs.registerListeners/onUpdated@moz-extension://ab6c7706-2274-4028-9b5b-c9dcc311be53/js/vapi-background.js:382:9
```

**Comment by gwarser** [comment]
https://www.mozilla.org/en-US/firefox/organizations/all/?q=English%20(US)

My Mint 18.3 KDE requires these libraries:

libgtk-3-0:i386
libdbus-glib-1-2:i386
libxt6:i386

**Comment by uBlock-user** [comment]
> Looks like you may be right.

You mean my bad 32 bit build claim ?

**Comment by gwarser** [comment]
No, I thought uBO uses WebAssembly with some 64bit code, but it not.

**Comment by gorhill** [comment]
The WebAssembly code was an experiment to find out if helped performance-wise. It did not but I kept the code around to revisit eventually to find out whether the overhead incurred when crossing the js/wasm barrier has been lowered.

**Comment by gwarser** [comment]
32bit build of 61.0.1 is OK.

**Comment by gorhill** [comment]
Possibly an issue with running Firefox 32-bit in a 64-bit linux?

I am running mint 64-bit, how do I get to install the missing dependencies?

**Comment by gwarser** [comment]
When I run `./firefox -P "esr-32bit-test" -no-remote -jsconsole` from `firefox` directory I see error with name of missing library, then:

`sudo apt install library:i386`

**Comment by sh-dvl** [comment]
first, thanks for your tests on sunday...

tested again: 
1. tried with all addons deactivated, all runs well, no error on stdout, following on jsconsole:
```
[Exception... "Component returned failure code: 0x80004004 (NS_ERROR_ABORT) [nsIPrefBranch.getComplexValue]"  nsresult: "0x80004004 (NS_ERROR_ABORT)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
[Exception... "Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getComplexValue]"  nsresult: "0x8000ffff (NS_ERROR_UNEXPECTED)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
```

2. with only uBO active, again high load and bad response (10s or more) to open menu, addons-page or anything else. various outputs:
```
1532881733362	addons.webextension.uBlock0@raymondhill.net	WARN	Loading extension 'uBlock0@raymondhill.net': Reading manifest: Error processing permissions.1: Unknown permission "privacy"
Extension error: TypeError: command.suggested_key is null chrome://browser/content/ext-commands.js 79
[[Exception stack
loadCommandsFromManifest@chrome://browser/content/ext-commands.js:79:11
CommandList@chrome://browser/content/ext-commands.js:24:19
@chrome://browser/content/ext-commands.js:222:30
runSafeSyncWithoutClone@resource://gre/modules/ExtensionUtils.jsm:71:14
emit/promises<@resource://gre/modules/ExtensionUtils.jsm:384:55
emit@resource://gre/modules/ExtensionUtils.jsm:383:20
runManifest@resource://gre/modules/Extension.jsm:733:23
startup/<@resource://gre/modules/Extension.jsm:813:14
Current stack
runSafeSyncWithoutClone@resource://gre/modules/ExtensionUtils.jsm:73:129
emit/promises<@resource://gre/modules/ExtensionUtils.jsm:384:55
emit@resource://gre/modules/ExtensionUtils.jsm:383:20
runManifest@resource://gre/modules/Extension.jsm:733:23
startup/<@resource://gre/modules/Extension.jsm:813:14
]]
Extension error: out of memory undefined undefined
[[Exception stack
undefinedCurrent stack
runSafeSyncWithoutClone@resource://gre/modules/ExtensionUtils.jsm:73:129
runSafeSync@resource://gre/modules/ExtensionUtils.jsm:104:37
runSafe@resource://gre/modules/ExtensionCommon.jsm:123:32
receiveMessage@resource://gre/modules/ExtensionChild.jsm:586:49
_callHandlers/<@resource://gre/modules/MessageChannel.jsm:574:17
_callHandlers@resource://gre/modules/MessageChannel.jsm:573:14
_handleMessage/deferred.promise<@resource://gre/modules/MessageChannel.jsm:637:7
_handleMessage@resource://gre/modules/MessageChannel.jsm:634:24
receiveMessage@resource://gre/modules/MessageChannel.jsm:165:5
]]
```

```
[Exception... "Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getComplexValue]"  nsresult: "0x8000ffff (NS_ERROR_UNEXPECTED)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Error: Invalid option requestBody  WebRequest.jsm:88:15
TypeError: commandsMap.get(...) is undefined[Weitere Informationen]  ext-commands.js:252:9
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
[Exception... "Component returned failure code: 0x80004004 (NS_ERROR_ABORT) [nsIPrefBranch.getComplexValue]"  nsresult: "0x80004004 (NS_ERROR_ABORT)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
1532883189548	addons.webextension.uBlock0@raymondhill.net	WARN	Loading extension 'uBlock0@raymondhill.net': Reading manifest: Error processing permissions.1: Unknown permission "privacy"
command.suggested_key is null  ext-commands.js:79
[Exception... "Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getComplexValue]"  nsresult: "0x8000ffff (NS_ERROR_UNEXPECTED)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Error: Invalid option requestBody  WebRequest.jsm:88:15
TypeError: commandsMap.get(...) is undefined[Weitere Informationen]  ext-commands.js:252:9
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
```

```
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Error: Invalid option requestBody  WebRequest.jsm:88:15
TypeError: commandsMap.get(...) is undefined[Weitere Informationen]  ext-commands.js:252:9
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
[Exception... "Component returned failure code: 0x80004004 (NS_ERROR_ABORT) [nsIPrefBranch.getComplexValue]"  nsresult: "0x80004004 (NS_ERROR_ABORT)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
1532883302460	addons.webextension.uBlock0@raymondhill.net	WARN	Loading extension 'uBlock0@raymondhill.net': Reading manifest: Error processing permissions.1: Unknown permission "privacy"
command.suggested_key is null  ext-commands.js:79
[Exception... "Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getComplexValue]"  nsresult: "0x8000ffff (NS_ERROR_UNEXPECTED)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Error: Invalid option requestBody  WebRequest.jsm:88:15
TypeError: commandsMap.get(...) is undefined[Weitere Informationen]  ext-commands.js:252:9
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIDOMWindowUtils.isParentWindowMainWidgetVisible]  nsPrompter.js:350
Error: Script terminated by timeout at:
HNTrieBuilder.prototype.vacuum@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/hntrie.js:333:13
HNTrieBuilder.fromDomainOpt@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/hntrie.js:85:12
FilterOriginHitSet.prototype<.matchOrigin.value@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:890:30
FilterOrigin.prototype.match@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:794:12
FilterBucket.prototype.match@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:1319:14
FilterContainer.prototype.matchTokens@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:2536:29
FilterContainer.prototype.matchStringGenericHide@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:2564:21
FilterContainer.prototype.matchStringExactType@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:2588:16
PageStore.prototype.init@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/pagestore.js:322:22
PageStore.prototype.reuse@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/pagestore.js:378:5
µb.bindTabToPageStats@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/tab.js:878:5
vAPI.tabs.onNavigation@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/tab.js:491:25
vAPI.tabs.registerListeners/onCommitted@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/vapi-background.js:364:9
  hntrie.js:333:13
Invalid match pattern: 'moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIDOMWindowUtils.isParentWindowMainWidgetVisible]  nsPrompter.js:350
Error: Script terminated by timeout at:
HNTrieBuilder.prototype.vacuum@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/hntrie.js:333:13
HNTrieBuilder.fromDomainOpt@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/hntrie.js:85:12
FilterOriginHitSet.prototype<.matchOrigin.value@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:890:30
FilterOrigin.prototype.match@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:794:12
FilterContainer.prototype.matchTokens@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:2527:33
FilterContainer.prototype.matchString@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/static-net-filtering.js:2765:22
PageStore.prototype.filterRequest@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/pagestore.js:640:18
onBeforeRequest@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/traffic.js:157:18
vAPI.net.registerListeners/onBeforeRequest@moz-extension://f5dc6456-d045-49f2-bfa8-f105cf573a48/js/vapi-webrequest.js:200:16
  hntrie.js:333:13
```

```
1532883302460	addons.webextension.uBlock0@raymondhill.net	WARN	Loading extension 'uBlock0@raymondhill.net': Reading manifest: Error processing permissions.1: Unknown permission "privacy"
Extension error: TypeError: command.suggested_key is null chrome://browser/content/ext-commands.js 79
[[Exception stack
loadCommandsFromManifest@chrome://browser/content/ext-commands.js:79:11
CommandList@chrome://browser/content/ext-commands.js:24:19
@chrome://browser/content/ext-commands.js:222:30
runSafeSyncWithoutClone@resource://gre/modules/ExtensionUtils.jsm:71:14
emit/promises<@resource://gre/modules/ExtensionUtils.jsm:384:55
emit@resource://gre/modules/ExtensionUtils.jsm:383:20
runManifest@resource://gre/modules/Extension.jsm:733:23
startup/<@resource://gre/modules/Extension.jsm:813:14
Current stack
runSafeSyncWithoutClone@resource://gre/modules/ExtensionUtils.jsm:73:129
emit/promises<@resource://gre/modules/ExtensionUtils.jsm:384:55
emit@resource://gre/modules/ExtensionUtils.jsm:383:20
runManifest@resource://gre/modules/Extension.jsm:733:23
startup/<@resource://gre/modules/Extension.jsm:813:14
]]
ExceptionHandler::GenerateDump cloned child 2601
ExceptionHandler::WaitForContinueSignal waiting for continue signal...
ExceptionHandler::SendContinueSignalToChild sent continue signal to child

###!!! [Parent][RunMessage] Error: Channel error: cannot send/recv


###!!! [Parent][RunMessage] Error: Channel error: cannot send/recv


###!!! [Parent][RunMessage] Error: Channel error: cannot send/recv


###!!! [Parent][RunMessage] Error: Channel error: cannot send/recv


###!!! [Parent][RunMessage] Error: Channel error: cannot send/recv
```

I'll try it today or tomorrow on a pure 32bit-machine to verify, if it's really a problem of 32bit on 64bit-machine

**Comment by sh-dvl** [comment]
sorry, no time yesterday.
on a linuxmint 18.3 32bit with firefox 52.9.0esr with uBO 1.16.12 (all other addons deactivated) after switching uBO active it slows down again and after some tries this occures:

```
[Exception... "Component returned failure code: 0x80004004 (NS_ERROR_ABORT) [nsIPrefBranch.getComplexValue]"  nsresult: "0x80004004 (NS_ERROR_ABORT)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
1533025405594	addons.webextension.uBlock0@raymondhill.net	WARN	Loading extension 'uBlock0@raymondhill.net': Reading manifest: Error processing permissions.1: Unknown permission "privacy"
command.suggested_key is null  ext-commands.js:79
Invalid match pattern: 'moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Error: Invalid option requestBody  WebRequest.jsm:88:15
TypeError: commandsMap.get(...) is undefined[Weitere Informationen]  ext-commands.js:252:9
Invalid match pattern: 'moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
[Exception... "Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getComplexValue]"  nsresult: "0x8000ffff (NS_ERROR_UNEXPECTED)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIDOMWindowUtils.isParentWindowMainWidgetVisible]  nsPrompter.js:350
Error: Script terminated by timeout at:
HNTrieBuilder.prototype.vacuum@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/hntrie.js:333:13
HNTrieBuilder.fromDomainOpt@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/hntrie.js:85:12
FilterOriginHitSet.prototype<.matchOrigin.value@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:890:30
FilterOrigin.prototype.match@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:794:12
FilterBucket.prototype.match@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:1319:14
FilterContainer.prototype.matchTokens@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:2536:29
FilterContainer.prototype.matchStringGenericHide@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:2564:21
FilterContainer.prototype.matchStringExactType@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:2588:16
PageStore.prototype.init@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/pagestore.js:322:22
PageStore@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/pagestore.js:244:5
PageStore.factory@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/pagestore.js:256:17
µb.bindTabToPageStats@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/tab.js:857:21
vAPI.tabs.onUpdated@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/tab.js:514:5
vAPI.tabs.registerListeners/onUpdated@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/vapi-background.js:382:9
  hntrie.js:333:13
Invalid match pattern: 'moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
```

then I have updated uBO (seen too late, that it's not up to date) and first all looks well, but then after a few stop/starts of firefox again:

```
[Exception... "Component returned failure code: 0x80004004 (NS_ERROR_ABORT) [nsIPrefBranch.getComplexValue]"  nsresult: "0x80004004 (NS_ERROR_ABORT)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
1533026101644	addons.webextension.uBlock0@raymondhill.net	WARN	Loading extension 'uBlock0@raymondhill.net': Reading manifest: Error processing permissions.1: Unknown permission "privacy"
command.suggested_key is null  ext-commands.js:79
Invalid match pattern: 'moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Error: Invalid option requestBody  WebRequest.jsm:88:15
TypeError: commandsMap.get(...) is undefined[Weitere Informationen]  ext-commands.js:252:9
Invalid match pattern: 'moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
[Exception... "Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getComplexValue]"  nsresult: "0x8000ffff (NS_ERROR_UNEXPECTED)"  location: "JS frame :: resource://app/components/nsBrowserContentHandler.js :: get startPage :: line 565"  data: no]  (unbekannt)
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIDOMWindowUtils.isParentWindowMainWidgetVisible]  nsPrompter.js:350
Error: Script terminated by timeout at:
HNTrieBuilder.prototype.vacuum@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/hntrie.js:333:13
HNTrieBuilder.fromDomainOpt@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/hntrie.js:85:12
FilterOriginHitSet.prototype<.matchOrigin.value@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:890:30
FilterOrigin.prototype.match@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:794:12
FilterBucket.prototype.match@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:1319:14
FilterContainer.prototype.matchTokens@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:2536:29
FilterContainer.prototype.matchStringGenericHide@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:2564:21
FilterContainer.prototype.matchStringExactType@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/static-net-filtering.js:2588:16
PageStore.prototype.init@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/pagestore.js:322:22
PageStore@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/pagestore.js:244:5
PageStore.factory@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/pagestore.js:256:17
µb.bindTabToPageStats@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/tab.js:857:21
vAPI.tabs.onUpdated@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/tab.js:514:5
vAPI.tabs.registerListeners/onUpdated@moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/js/vapi-background.js:382:9
  hntrie.js:333:13
Invalid match pattern: 'moz-extension://7fe93410-06ac-4655-8716-8ac3853cb3ab/web_accessible_resources/*'  MatchPattern.jsm:52
Invalid match pattern: 'ws://*/*'  MatchPattern.jsm:52
Invalid match pattern: 'wss://*/*'  MatchPattern.jsm:52
Content Security Policy: Unbekannte Direktive 'worker-src' kann nicht verarbeitet werden  (unbekannt)
Content Security Policy: Die Direktive 'frame-src'  sollte nicht mehr verwendet werden. Bitte verwenden Sie stattdessen die Direktive 'child-src'.  (unbekannt)
TypeError: can't access dead object  webbrowser.js:2508:1
TypeError: can't access dead object  webbrowser.js:2508:1
TypeError: can't access dead object  webbrowser.js:2508:1
```

well, it looks like that this is a not 100% deterministic problem with uBO connecting to firefox-32bit-API and using it on a 64bit-system it seems to get nearly to the 100%.

as you noted above, uBO is javascript-only, so it makes no difference between architecture, but at least it can trigger this behaviour, so maybe it can be determined, what have to occure for trigger, so it can be fixed then in FF-API?

**Comment by uBlock-user** [comment]
This appears more and more of a specific broken build of Firefox ESR, you cannot fix a broken build by submitting patches.

**Comment by sh-dvl** [comment]
might be, but it's not compiled by myself, but it's this: http://ftp.mozilla.org/pub/thunderbird/releases/52.9.0/linux-i686/de/thunderbird-52.9.0.tar.bz2

**Comment by uBlock-user** [comment]
If Mozilla still supports entertaining bugs from legacy builds, you should file this bug there and let them know about this situation.

**Comment by sh-dvl** [comment]
mmmh, and because it only occures with uBO, they then say: file a bug here...

**Comment by uBlock-user** [comment]
> file a bug here...

and you already filed it here, so just reference this one there.

**Comment by sh-dvl** [comment]
sorry, but maybe you didn't get the point: if no-one feels responsible, nothing will happen...

if something gets wrong with a program using extensions and I can reproducibly trigger it with exactly one certain extension, but no other, the normal way is to contact the maintainer of the extension and not of the program. 

If then a possibility is found, to trigger the problem with either extension or anything else, then the ball is at the maintainer of the product.

**Comment by uBlock-user** [comment]
> if no-one feels responsible, nothing will happen...

Errors as above thrown by the browser itself are browser devs responsibility, extension maintainers cannot fix something which is already thoroughly messed up in the browser itself. 

> I can reproducibly trigger it with exactly one certain extension, but no other, the normal way is to contact the maintainer of the extension and not of the program.

I would suggest you try repoducing this issue on newer builds like v54, v55, v56, this will further help eliminate the possibility of this being a uBO bug as you believe.

**Comment by gorhill** [comment]
I will need to investigate before an issue is opened on bugzilla, no point dumping this in there until the issue is investigated.

**Comment by uBlock-user** [comment]
In regards to `TypeError: can't access dead object  webbrowser.js:2508:1` there was a bug filed in [bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1409705), could this be related to that ?

**Comment by gorhill** [comment]
I can't remember the name of the setting now, but there is something in `about:config` where we can adjust the timeout value for when javascript execute. I remember I had to advise someone a while ago to maybe increase this value because he was using old hardware and this was causing uBO to time out in some place, and on older version of Firefox, the timeout was cascading into more issues because Firefox was trying to show the prompter from the background script (which is what you see in the call stacks above).

So my guess is that there is a performance issue with Firefox 32-bit + older hardware (?) causing this issue. The vacuum method can be expensive for some of the very long  `domain=` options in EasyList, and this may be worse for some older Firefox -- I always benchmarked using current Firefox. Though the vacuum method itself is known to be CPU-intensive, it's lazily called only once for a given `domain=` option and once its done, the benefit is faster domain lookups when the `domain=` option has to be evaluated afterward.

**Comment by gorhill** [comment]
@sh-dvl Something to experiment: what results do you obtain on [that benchmark](https://gorhill.github.io/obj-vs-set-vs-map/set-vs-regexp.html) for the "Create large-size dictionary / Trie-based" entry? I get this on my side with Nitghtly:

    Create large-size dictionary
    - Set-based x 7,076 ops/sec ±5.36% (47 runs sampled)
    - Regex-based x 9,944 ops/sec ±3.69% (53 runs sampled)
    - Trie-based x 1,312 ops/sec ±2.58% (58 runs sampled) Done.

"Trie-based" is essentially the `vacuum` method.

**Comment by sh-dvl** [comment]
can currently only try on pure 32bit-system, but for "Tries-based" I have to increase `dom.max_script_run_time`, else firefox warn for slow-down and if not clicked `wait` within some seconds, firefox seems to stop it without no further notice.

after increase there still nothing happen after some minutes, but on java-console, there comes a 
`uncaught exception: out of memory  (unbekannt)`

seeing this, I started a ps + a free within a loop, calling them every second, and near that OOM it says:
```
              total        used        free      shared  buff/cache   available
Mem:        2062324      381492     1404868       46576      275964     1402696
Swap:       4193276      545980     3647296
0  1000 20071 16925  20   0 808068 224048 poll_s Sl+ pts/2      0:03 /opt/firefox/firefox -jsconsole
0  1000 20114 20071  20   0 734932 102188 poll_s Sl+ pts/2      1:19 /opt/firefox-52.9.0esr/plugin-container -greomni /opt/firefox-52.9.0esr/omni.ja -appomni /opt/firefox-52.9.0esr/browser/omni.ja -appdir /opt/firefox-52.9.0esr/browser 20071 true tab
```

so there should be no OOM in firefox, but maybe firefox have a self-defined limit?

output until then with no addons active (even that won't really help you):
```
Benchmarking, the higher ops/sec the better.
Firefox 52.0 on Linux i686.

Create small-size dictionary
  -   Set-based x 811,852 ops/sec ±7.28% (39 runs sampled)
  - Regex-based x 443,424 ops/sec ±3.76% (39 runs sampled)
```

if you've suggestions, how to try this in another way, I'll try it, but for the next hours I can't due to other undertakings

**Comment by gwarser** [comment]
`dom.max_ext_content_script_run_time` was introduced in Fx 57 https://bugzilla.mozilla.org/show_bug.cgi?id=1356334

**Comment by gorhill** [comment]
I stepped into the code and all is working fine, i.e. as expected. However when let the benchmark run, I see memory going up fast in system monitor -- to the point where I have to abort the script or the OS will end up paging, causing full desktop freeze etc. So something appears to be wrong with Firefox's 52 TypedArray on the 32-bit version -- possibly only a linux issue given that there is no widespread reports of the issue here. Will need to look at TypedArray issues fixed on bugzilla to find out if something was fixed which we can relate to the issue here.

**Comment by gorhill** [comment]
To be sure nobody waste their time while I keep investigating, this has nothing to do with TypedArray.

**Comment by gwarser** [comment]
I tried "Lookup" benchmark in 52 32bit and it works on first run after browser startup, but timeouts on medium dict when I try to run it again.

I think esr 60 32bits auto updates to 64bits.

**Comment by gorhill** [comment]
At this point it does look like an issue with Firefox.

I added a `debugger` statement to be taken whenever the `vacuum` seems to take longer than it should. It is being taken, and I can see that the input array, which is supposed to contain only integer numbers has a value like `1.11...e-311` in it. The exact value is random but it's always something very small, like `e-[160-312]`.

I added code to detect such number at array filling time, in the `add()` method, and such number is not being detected.

I also added code to detect bad values in the array at the top of `vacuum()`, and it's detected randomly. Sometime even when there is nothing bad found in the array at the top of `vacuum`, something bad is found during processing of the array.

This suggests that for some reasons the content of the array is mutated internally after `add()` and/or before or during `vacuum()` and this causes the issue. At this point this is confirmed to be an issue specific to Firefox 52 ESR 32-bit on Linux.

**Comment by sh-dvl** [comment]
well, many thanks to all of you, who helped to isolate possible cause of this problem.

this evening I had the chance to test on a single-user i7-8xxx with 32GB RAM (so no performance-issue should occure) with an 32bit 52.8.0esr-firefox: no problem. even the benchmark runs well.

using a 32bit 52.9.0esr, the same issue as described occures.

using a 64bit 52.9.0esr again no problem, even with the benchmark

I've now filed a bug at https://bugzilla.mozilla.org/show_bug.cgi?id=1479907
